### ISSUE TITLE
rough implementation doubt

### ISSUE THREAD

**Issue by vainaijr** [pytorch/pytorch#34760]
I created a rough implementation of stand alone self attention, I have certain doubts regarding it, should I create a pull request with [WIP], to iterate and figure out solutions to mistakes in my implementation.

**Comment by VitalyFedyunin** [comment]
CC @zhangguanheng66

**Comment by VitalyFedyunin** [comment]
Hi @zhangguanheng66 can you please act as domain expert here

**Comment by zhangguanheng66** [comment]
Do you mean the self-attention building blocks in MutilheadAttention module? Something like this
https://github.com/pytorch/pytorch/pull/34793

**Comment by vainaijr** [comment]
I do not modify existing MHA, in the paper for stand alone self attention, they use notion of relative distance between pixels, so how far a pixel is to the query will also have an impact on attention weights, 

nn.MHA only expect input as text sequence, that is q, k, v is only 3 dimensional tensor, (sequence length, batch size, embedding size), but for stand alone self attention, q, k, v is 4 dimensional tensor, (batch size, channels, height, width).

So, instead of nn.Linear, nn.Conv2d() with kernel_size 1, is used to obtain q, k, v, and then unfolding is done, to obtain kernel size key, value, then compute relative distance embedding, and the formula in stand alone self attention is modified, that is softmax(q*(k + rel_distance))*v, as we consider impact of how far query pixel is to key pixels also.

this work a bit different than existing MHA, because of image input, and relative distance notion.

**Comment by zhangguanheng66** [comment]
For the input dimension part, `ScaledDotProduct` module in https://github.com/pytorch/pytorch/pull/34793 accepts tensors with more than 3 dimension  because `nn.matmul` is applied. And should the dimension be (N, H, W, C)?

For the attention layer with the consideration of `rel_distance`, IMO, it's a little bit too custom for a platform. Pulling specific implementations into the library limits the rich, on-going platform innovation.

**Comment by vainaijr** [comment]
Yes, it would be (N, H, W, C);  they use different way to compute rel distance in different paper, so, it is a bit custom. For example, in this paper, https://openreview.net/pdf?id=ryfxVNEajm, they use skewing technique to compute relative distance.

I think only making input as >3 dimensional would be ok for now, that is all three of q, k, v could be >3 dimensional.
Also, instead of using bmm or nn.matmul, is einsum a better way to do matrix multiplication?
